### PR TITLE
ARUHA-406: Avoid Kafka unavailability from cascading to Nakadi

### PR DESCRIPTION
**Pull request by adyach**
This change only for publishing events.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -145,6 +147,7 @@ dependencies {
         exclude module: "json"
     }
     compile 'com.fasterxml.jackson.datatype:jackson-datatype-joda:2.4.0'
+//    compile 'com.netflix.hystrix:hystrix-core:1.5.6'
```
Comment (v1ctor) [inline build.gradle]: I think you have to remove it then?

Comment (adyach) [inline build.gradle]: 7d2995b

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Command is aimed to prevent calls in cases Kafka is not available.
+ */
+public class KafkaSyncPostCommand extends HystrixCommand<KafkaSyncPostCommand.CommandResult> {
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: What exactly Hystrix will do here?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaSyncPostCommand.class);
+    private static final String GROUP_KEY_SUFFIX = "kafka-sync-post-";
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: Is it s suffix or a prefix?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    private static final int HYSTRIX_COMMAND_TIMEOUT_DELTA = 5000;
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: Isn't it better to extract these configurations to application yml?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            if (hasKafkaConnectionException(exceptions)) {
+                throw eventPublishingException;
+            }
+            return new CommandResult(eventPublishingException);
+        }
+        return new CommandResult();
+    }
+
+    private boolean hasKafkaConnectionException(final List<Exception> exceptions) {
+        return exceptions.stream().filter(ex ->
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: What about `anyMatch`?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        } catch (final RuntimeException e) {
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        }
+    }
+
+    public static class CommandResult {
+
+        private final EventPublishingException exception;
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: What about using subclassing to differentiate between a successful command from a failed one? We could have abstract CommandResult and then SuccessfulCommandResult and FailedCommandResult. What do you think?

Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: @rcillo I thought about it but there are no so many cases for that. It is very specific case when we have to return one exception and throw another. I think it is easier to understand with less code.

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java (hunk 5 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: Thanks, fixed 5a403ff

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java (hunk 4 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: Yep, you are right, thanks 5a403ff

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java (hunk 3 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommand.java]: fixed 5a403ff

File: src/test/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Test
+    public void testUnknownTopicOrPartitionException() {
+        final BatchItem batchItem = new BatchItem(new JSONObject());
+        batchItem.setPartition("0");
+        mockedKafkaProducer.setException(new UnknownTopicOrPartitionException());
+
+        final KafkaSyncPostCommand.CommandResult result = createCommand(batchItem).execute();
+
+        Assert.assertTrue(result.isFailed());
```
Comment (rcillo) [inline src/test/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommandTest.java]: When does hystrix increase the failure counter? Is it when an exception is launched in the `execute` command?

Comment (rcillo) [inline src/test/java/org/zalando/nakadi/repository/kafka/KafkaSyncPostCommandTest.java]: Ok, got it. It considers a failure when an exception is launched during execution.

**Comment by rcillo** [comment]
As we discussed previously, we have some concerns on the impact of this feature since it fails entire topics, even if they have many partitions in different brokers.

We analysed the possibility to group errors on a per broker basis, but it seems to require a bigger effort. We also considered grouping per topic-partition which is also hard to implement.

So I would like to test how it behaves first to make sure we already got some improvement over the status quo, which is dramatically failing everything.

**Comment by rcillo** [comment]
:shipit:

**Comment by adyach** [comment]
@rcillo I've fixed tests parameters, pls, approve again.

**Comment by rcillo** [comment]
:shipit:

**Pull request merged**